src/lib/operators/jit_operator/jit_types.hpp

```diff
@@ -0,0 +1,142 @@
+ * operations with specialized versions for the concrete data types.
+ *
+ * This entails two important restrictions:
+ * 1) There can be no temporary (=local) variant values. This is because we support std::string as a data type.
+ *    Strings are not POD types and they thus require memory allocations, destructors and exception
+ *    handling. The specialization engine is not able to completely optimize this overhead away, even if no strings are
+ *    processed in an operation.
+ * 2) The data type of each value needs to be stored separately from the value itself. This is because we want the
+ *    specialization engine to know about data type, but not concrete values.
+ *
+ * Both restrictions are enforced using the same mechanism:
+ * All values (and other data that must not be available to the specialization engine) are encapsulated in the
+ * JitRuntimeContext. This context is only passed to the operator after code specialization has taken place.
+ * The JitRuntimeContext contains a vector of variant values (called tuple) which the JitOperator can access through
+ * JitTupleValue instances.
+ * The runtime tuple is created and destroyed outside the JitOperator's scope, so no memory management is
+ * required from within the operator.
+ * Each JitTupleValue encapsulates information about how to access a single value in the runtime tuple. However, the
+ * JitTupleValues are part of the JitOperator and must thus not store a reference to the JitRuntimeContext. So while
+ * they "know" how to access values in the runtime tuple, they do not have the means to do so.
+ * Only by passing the runtime context to a JitTupleValue, a JitMaterializedValue is created.
+ * This materialized value finally allows the operator to access a data value.
+ */
+
+/* Since we handle arithmetic and logical operations in the same way inside the JitOperator,
+ * we need an additional boolean type. Booleans are stored as uint8_t, since std::vector<bool>
+ * is implemented in space-efficient manner in most standard library implementations and does not support
+ * all necessary operations (e.g. accessing elements as non-const references).
+ * Null, on the other hand, is not a valid data type. Instead, each value must have a (potential) concrete data type.
+ * Nullable values have an additional boolean value indicating their NULL status.
+ */
+#define JIT_DATA_TYPE_INFO ((uint8_t, Bool, "bool")) DATA_TYPE_INFO
+
+enum class JitDataType : uint8_t { Bool, BOOST_PP_SEQ_ENUM(DATA_TYPE_ENUM_VALUES) };
+
+// This templated method returns the JitDataType for a given type.
+template <typename T>
+JitDataType jit_data_type();
+
+// A vector of variant values.
+// In order to store N variant values, one vector of size N is created per data type.
+// Each of the N variant values is then represented by the elements at some index I in each vector.
+class JitVariantVector {
```

Comment:
Does this resemble a row?

Reply:
Could you plase describe the relationships between the types defined in this class?